shell/platform/tizen/tizen_vsync_waiter.h

```diff
@@ -5,41 +5,35 @@
 #ifndef EMBEDDER_TIZEN_VSYNC_WAITER_H_
 #define EMBEDDER_TIZEN_VSYNC_WAITER_H_
 
-#include <Ecore.h>
 #include <tdm_client.h>
 
 #include <thread>
 
 #include "flutter/shell/platform/embedder/embedder.h"
 
+class TizenEmbedderEngine;
+
 class TizenVsyncWaiter {
  public:
-  TizenVsyncWaiter();
+  TizenVsyncWaiter(TizenEmbedderEngine* engine);
   virtual ~TizenVsyncWaiter();
-  bool CreateTDMVblank();
-  bool AsyncWaitForVsync();
   void AsyncWaitForVsync(intptr_t baton);
-  void AsyncWaitForRunEngineSuccess(FLUTTER_API_SYMBOL(FlutterEngine)
-                                        flutter_engine);
+
+  bool IsValid();
 
  private:
-  static const int VBLANK_LOOP_REQUEST = 1;
-  static const int VBLANK_LOOP_DEL_PIPE = 2;
-  void AsyncWaitForVsyncCallback();
-  void DeleteVblankEventPipe();
-  static void CreateVblankEventLoop(void* data);
+  bool CreateTDMVblank();
+  void HandleVblankLoopRequest();
   static void TdmClientVblankCallback(tdm_client_vblank* vblank,
                                       tdm_error error, unsigned int sequence,
                                       unsigned int tv_sec, unsigned int tv_usec,
                                       void* user_data);
-  static void VblankEventLoopCallback(void* data, void* buffer,
-                                      unsigned int nbyte);
-  tdm_client* client_;
-  tdm_client_output* output_;
-  tdm_client_vblank* vblank_;
-  FLUTTER_API_SYMBOL(FlutterEngine) flutter_engine_;
-  intptr_t baton_;
-  Ecore_Pipe* vblank_ecore_pipe_;
+
+  tdm_client* client_{nullptr};
+  tdm_client_output* output_{nullptr};
+  tdm_client_vblank* vblank_{nullptr};
+  TizenEmbedderEngine* engine_{nullptr};
```

Comment:
I think it would be better than before.

First of all, we run ```FlutterEngineRun``` to initialize ```flutter_engine``` in ```TizenEmbedderEngine::RunEngine```
But ```TizenEmbedderEngine::OnVsyncCallback``` is called even before FlutterEngineRun call is over(In fact, this was the cause of many problems).

Therefore, at that time, we cannot set flutter_engine as a TizenVsyncWaiter member in main thread flow.(However, Actually, the flutter_engine is already initialized that moment) 

Of course when ```OnVsyncCallback``` is first called, we can set flutter_engine as an argument of ```TizenVsyncWaiter::AsyncWaitForVsync``` or we can set  as a member of ```TizenVsyncWaiter```.

These are also appropriate solutions.  But I thought that having ```TizenEmbedderEngine``` as a member would be a better solution.

Reply:
Because of this context, we were able to remove below. 
```
tizen_vsync_waiter_->AsyncWaitForRunEngineSuccess(flutter_engine); // In most cases, the baton was zero 
```

Reply:
we need conisder when OnVsyncCallback first time. At this time engine maybe not created.

Reply:
I considered it and I judge it early by ```IsValid```.
```TizenVsyncWaiter::TdmClientVblankCallback```(After all, OnVsyncCallback is for this.) will only work fine when all of related things are fine. 
if there is case with uninitialized factor, It will lead directly to a crash.

Reply:
I think that the point is that elimination of abnormal calling ``TizenVsyncWaiter::TdmClientVblankCallback```

Reply:
if call OnVsyncCallback first time and tizen_vsync_waiter_->IsValid == false,The flutter engine will not request vsync again.
I'ts  difficult to reproduce it, we need consider it.
about this issue, I have two solutions:
1.we can check whether need to request vsync after run flutter engine.
2.While the call to `FlutterEngineOnVsync` must occur on the thread that made the call to `FlutterEngineRun`

Reply:
> if call OnVsyncCallback first time and tizen_vsync_waiter_->IsValid == false,The flutter engine will not request vsync again. 

Well... I think this will never happen. 

Calling ```OnVsyncCallback``` is only possible after the shell is created, and the shell is created in FlutterEngineRunInitialized.
The flutter_engine's memory was already allocated by FlutterEngineInitialize at that time.
OnVsyncCallback is only called from another thread after the shell is created.

In this change, TizenVsyncWaiter has TizenEmbedderEngine as a member before shell creation and thread execution.
So I think IsValid should always be true because ```TizenVsyncWaiter``` accesses ```flutter_engine``` through ```TizenEmbedderEngine```

If the your concern is reproducible, could you please tell me how? If you let me know, I'll try to reproduce.

``` c++
FlutterEngineResult FlutterEngineRun(size_t version,
                                     const FlutterRendererConfig* config,
                                     const FlutterProjectArgs* args,
                                     void* user_data,
                                     FLUTTER_API_SYMBOL(FlutterEngine) *
                                         engine_out) {
  auto result =
      FlutterEngineInitialize(version, config, args, user_data, engine_out); // initialize flutter_engine
  if (result != kSuccess) {
    return result;
  }

  return FlutterEngineRunInitialized(*engine_out); // create shell and run thread 
```

Reply:
yes, you are right~

Reply:
Thank you for review, I'll resolve this